File: leetcode/problems/450.ts
function deleteNode(root: TreeNode | null, key: number): TreeNode | null {
  if (!root) {
    return root
  }
  if (key > root.val) root.right = deleteNode(root.right, key)
  else if (key < root.val) root.left = deleteNode(root.left, key)
  else {
    if (!root.left) root = root.right
    else if (!root.right) root = root.left
    else {
      let t = root.right
      while (t.left) {
        t = t.left
      }
      t.left = root.left
      root = root.right
    }
  }
  return root
}

File: leetcode/problems/450.go
func deleteNode(root *TreeNode, key int) *TreeNode {
  if root == nil {
    return nil
  }
  if key > root.Val {
    root.Right = deleteNode(root.Right, key)
  } else if key < root.Val {
    root.Left = deleteNode(root.Left, key)
  } else {
    if root.Left == nil {
      root = root.Right
    } else if root.Right == nil {
      root = root.Left
    } else {
      cur := root.Right
      for cur.Left != nil {
        cur = cur.Left
      }
      cur.Left = root.Left
      root = root.Right
    }
  }
  return root
}